### PR TITLE
Convert GLO ephemeris to SBP

### PR DESCRIPTION
**Pull request by mfine**
Based on [peddie/glonass](https://github.com/swift-nav/gnss-converters/compare/peddie/glonass).

Converts GLO ephemerides messages in RTCM to SBP.

/cc @peddie @anth-cole @benjamin0

### CODE REVIEW

**Comment by mfine** [comment]
Outstanding issues are labeled with TODO:

+ default value for fit interval
+ d tau field mapping
+ fcn mapping
+ iod

Also, just generally does the conversion look any good.

**Comment by benjamin0** [comment]
cc @scarcanague

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs [hunk 1]
```diff
@@ -75,7 +113,18 @@ fitInterval fitInt iodc
...
 
+-- | Convert GLONASS fit interval - Table 4.3 from GLONASS ICD.
+--
+glonassFitInterval :: Word8 -> Word32
+glonassFitInterval fi
+  | fi == 0 = 0
+  | fi == 1 = 30 * 60
+  | fi == 2 = 45 * 60
+  | fi == 3 = 60 * 60
+  | otherwise = 60 * 60 -- ?? TODO: GOOD IDEA? WHAT'S A GOOD DEFAULT?
```
Comment (anth-cole) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: I'd do the default fit interval to be ~1hr, after this the orbit errors are probably getting to large

**Comment by anth-cole** [comment]
conversion looks ok to me, I'd definitely need the last 3 fields before we deploy

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs [hunk 2]
```diff
@@ -85,46 +134,96 @@ toGpsEphemerisCommonContent m = do
...
+                 [ m ^. msg1020_ephemeris ^. glonassEphemeris_xndot
+                 , m ^. msg1020_ephemeris ^. glonassEphemeris_yndot
+                 , m ^. msg1020_ephemeris ^. glonassEphemeris_zndot
+                 ]
+             , _msgEphemerisGlo_acc    = (* 1000) . ((-30) ##) <$>
+                 [ m ^. msg1020_ephemeris ^. glonassEphemeris_xndotdot
+                 , m ^. msg1020_ephemeris ^. glonassEphemeris_yndotdot
+                 , m ^. msg1020_ephemeris ^. glonassEphemeris_zndotdot
+                 ]
+             , _msgEphemerisGlo_fcn    = m ^. msg1020_header ^. glonassEphemerisHeader_channel -- TODO: docs say FCN+8???
```
Comment (benjamin0) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: which docs?

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs [hunk 3]
```diff
@@ -85,46 +134,96 @@ toGpsEphemerisCommonContent m = do
...
+             , _msgEphemerisGlo_iod    = undefined -- TODO: ??? Issue of ephemeris data??
```
Comment (benjamin0) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: Is this not found in the ephemeris?

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs [hunk 4]
```diff
@@ -85,46 +134,96 @@ toGpsEphemerisCommonContent m = do
...
+-- | Convert an RTCM 1020 GLONASS ephemeris message into an SBP MsgEphemerisGlo.
+--
+glonassConverter :: MonadStore e m => Msg1020 -> Conduit i m [SBPMsg]
+glonassConverter m = do
+  common <- toGlonassEphemerisCommonContent m
+  let m' = MsgEphemerisGlo
+             { _msgEphemerisGlo_common = common
+             , _msgEphemerisGlo_gamma  = (-40) ## (m ^. msg1020_ephemeris ^. glonassEphemeris_gammaN)
+             , _msgEphemerisGlo_tau    = (-30) ## (m ^. msg1020_ephemeris ^. glonassEphemeris_tauN)
+             , _msgEphemerisGlo_d_tau  = undefined -- TODO: equipment delay between L1 and L2??
```
Comment (benjamin0) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: Is this also not found in the ephemeris? I thought it was.

Comment (scarcanague) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: It should correspond to field DF125 in RTCM

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs (hunk 1 above)
Comment (scarcanague) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: In Piksi firmware, we seem to default to 1 hour when the fit interval has not been decoded yet:
https://github.com/swift-nav/piksi_firmware_private/blob/c1be4b58e21434c47d1d45d4998e42f8fff5c797/src/nav_msg/nav_msg_glo.c#L433

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs (hunk 3 above)
Comment (scarcanague) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: The IOD is deduced from the 7 LSB of the `Tb`:
https://github.com/swift-nav/piksi_firmware_private/blob/c1be4b58e21434c47d1d45d4998e42f8fff5c797/src/nav_msg/nav_msg_glo.c#L521

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs (hunk 2 above)
Comment (scarcanague) [inline haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: The "-7" channel is 0 for RTCM while it is 1 for sbp

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs (hunk 1 above)
Comment (mfine) [reply haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: Thanks - will leave as is at `60 * 60`.

File: haskell/src/Data/RTCM3/SBP/Ephemerides.hs (hunk 4 above)
Comment (mfine) [reply haskell/src/Data/RTCM3/SBP/Ephemerides.hs]: Thanks! Picked up.

**Comment by mfine** [comment]
Review feedback picked up in 756164c.

**Comment by mfine** [comment]
First look at data...

from piksi:
```json
{
  "length": 120,
  "gamma": 0,
  "vel": [
    -88.36078643798828,
    157.0577621459961,
    3575.286865234375
  ],
  "iod": 12,
  "d_tau": 5.587935447692871e-09,
  "pos": [
    -18802502.44140625,
    -17240391.6015625,
    306566.89453125
  ],
  "common": {
    "health_bits": 0,
    "fit_interval": 2400,
    "valid": 1,
    "sid": {
      "sat": 14,
      "code": 3
    },
    "toe": {
      "wn": 1983,
      "tow": 251118
    },
    "ura": 2
  },
  "tau": -2.37245112657547e-05,
  "fcn": 1,
  "acc": [
    0,
    0,
    -9.313225746154784e-07
  ]
}
```

from converter:
```json
{
  "length": 120,
  "gamma": 0,
  "vel": [
    -7911.639213562012,
    157.0577621459961,
    3575.286865234375
  ],
  "iod": 12,
  "d_tau": 0,
  "pos": [
    -13965497.55859375,
    -15527608.3984375,
    306566.89453125
  ],
  "common": {
    "health_bits": 0,
    "fit_interval": 1800,
    "valid": 0,
    "sid": {
      "sat": 14,
      "code": 3
    },
    "toe": {
      "wn": 1983,
      "tow": 248418
    },
    "ura": 1
  },
  "tau": -0.001929400488734245,
  "fcn": 1,
  "acc": [
    0,
    0,
    -1.396983861923217e-05
  ]
}
```

**Comment by benjamin0** [comment]
@mfine looks like those are from two different ToE 45 minutes apart which might explain the different values.

**Comment by mfine** [comment]
@benjamin0 yeah, I'm trying to figure out if the ephemeris service is just behind.

**Comment by mfine** [comment]
Made some a bunch of fixes and seeing parity with outcoming ephemeris. Potentially there are some remaining issues. Going to bring this in for now. Thanks for the review!

**Pull request merged**